server/controllers/map-controllers.js

```diff
@@ -1,160 +1,79 @@
-        await user.save({ session: sess });
-        await sess.commitTransaction();
-        } catch (err) {
-        const error = new HttpError(
-            'Creating place failed, please try again.',
-            500
-        );
-        return next(error);
-    }
-
-    res.status(201).json({ place: createdPlace });
-}
-
-
-//getPlacesByUserId
-const getPlacesByUserId = async (req, res, next) => {
-    const userId = req.params.uid;
-  
-    // let places;
-    let userWithPlaces;
-    try {
-      userWithPlaces = await User.findById(userId).populate('places');
-    } catch (err) {
-      const error = new HttpError(
-        'Fetching places failed, please try again later.',
-        500
-      );
-      return next(error);
-    }
-  
-    if (!userWithPlaces || userWithPlaces.places.length === 0) {
-      return next(
-        new HttpError('Could not find places for the provided user id.', 404)
-      );
-    }
-  
-    res.json({
-      places: userWithPlaces.places.map(place =>
-        place.toObject({ getters: true })
-      )
-    });
+const getSafetyScore = require('../util/apis/safety');
+const getCoordsForAddress = require('../util/apis/location');
+const getTrafficInfo = require('../util/apis/traffic');
+
+const mapUtil = require('../util/map');
+
+// Only req, res, next, other util
+const getSearchByAddress = async (req, res, next) => {
+	// 1a. Get lat, lon from address
+	const { address } = req.body;
+
+	let coordinates;
+	try {
+		coordinates = await getCoordsForAddress(address);
+	} catch (error) {
+		console.error(error);
+		return;
+	}
+	// 1b. Check if place exist:
+	let place;
+	try {
+		place = await mapUtil.getPlaceByAddress(address);
+	} catch (error) {
+		return next(error);
+	}
+	// 1c. Save place
+	if (!place || place.length === 0) {
+		try {
+			place = await mapUtil.createPlace(address, coordinates);
+		} catch (error) {
+			return next(error);
+		}
+	}
+
+	// 2. Get Traffic documents using GeoSearch
+	let traffic;
+	try {
+		traffic = await mapUtil.getTrafficByLocation(coordinates);
+	} catch (error) {
+		return next(error);
+	}
+
+	// 3a. Check if no doc return, then make req to api
+	//neu traffic la array thi fai check if traffic.length ===0 nha
+	if (!traffic || traffic.length === 0) {
+		let newTrafficInfo = await getTrafficInfo(coordinates);
+
+		// 3b. Save traffic (mongo)
+		traffic = await mapUtil.createTraffic(newTrafficInfo);
+	}
+
+	// 4.Get SafetyByLocation in DB
+	let safetyScore;
+	try {
+		safetyScore = await mapUtil.getSafetyByLocation(coordinates);
+		// console.log('safety is ', safetyScore);
+	} catch (err) {
+		return next(err);
+	}
+
+	// 5a. Check if no doc return, then make req to api
+	if (safetyScore.length === 0) {
+		let newSafetyScore = await getSafetyScore(coordinates);
+
+		// 5b. Save safetyScore (mongo)
+		safetyScore = await mapUtil.createSafety(newSafetyScore);
+	}
+
+	// 6. Respond Object (2 key: safetyScore, traffic)
+	res.status(201).json({
+		Place: place,
```

Comment:
fix not bug de return place object di nha

Reply:
Fix roi ma 
<img width="1020" alt="Screen Shot 2023-06-03 at 9 49 53 PM" src="https://github.com/ankhuetang/TravelSafety/assets/88098962/898dc63f-035c-42f2-a302-e160d0128008">